src/epics/repositories.ts

```diff
@@ -0,0 +1,63 @@
+import { get } from '../fetcher';
+import { combineEpics } from 'redux-observable';
+import { Observable } from 'rxjs/Observable';
+import { Range, List } from 'immutable';
+import {
+  FETCH_USER_REPOS,
+  FETCH_ALL_REPOS,
+  FETCH_TOTAL_REPO_STARRED
+} from '../constants/repos';
+import { AddRepos } from '../actions/repositories';
+import { append } from '../actions/user';
+
+const fetchReposEpic = (action$) => (
+  action$
+    .ofType(FETCH_USER_REPOS)
+    .flatMap(({ username, page }) =>
+      get({
+        endpoint: `users/${username}/starred`,
+        params: { page }
+      })
+    )
+    .map(AddRepos)
+);
+
+const fetchTotalReposLengthEpic = (action$) => (
+  action$
+    .ofType(FETCH_TOTAL_REPO_STARRED)
+    .flatMap(({ username }) => (
+      get({
+        endpoint: `users/${username}/starred`,
+        params: { per_page: 1 },
+        resHeader: true
+      })
+    ))
+    .map(headers => {
+      const reg = /rel="next", <.*&page=(\d+)>; rel="last"/i;
+      const len = parseInt(headers.get('link').match(reg)[1], 10);
+      return append('starred', len);
+    })
+);
```

Comment:
Furthering our discussion on twitter, I usually suggest keeping these ajax + map stuff _inside_ your `mergeMap/concatMap/switchMap` (aka flatMap/flatMapLatest) because when/if someone ever adds error handling, they may unknowningly introduce bugs unless they move it.

Two examples:

```js
const fetchTotalReposLengthEpic = (action$) => (
  action$
    .ofType(FETCH_TOTAL_REPO_STARRED)
    .flatMap(({ username }) => (
      get({
        endpoint: `users/${username}/starred`,
        params: { per_page: 1 },
        resHeader: true
      })
      .catch(e => Observable.of({
        type: SOME_ERROR_TYPE,
        payload: e,
        error: true
      }))
    ))
    // this is now mapping not just the happy/success path
    // but also *incorrectly* mapping th error action from
    // our .catch() operator!
    .map(headers => {
      const reg = /rel="next", <.*&page=(\d+)>; rel="last"/i;
      const len = parseInt(headers.get('link').match(reg)[1], 10);
      return append('starred', len);
    })
);
```

const fetchTotalReposLengthEpic = (action$) => (
  action$
    .ofType(FETCH_TOTAL_REPO_STARRED)
    .flatMap(({ username }) => (
      get({
        endpoint: `users/${username}/starred`,
        params: { per_page: 1 },
        resHeader: true
      })
    ))
    .map(headers => {
      const reg = /rel="next", <.*&page=(\d+)>; rel="last"/i;
      const len = parseInt(headers.get('link').match(reg)[1], 10);
      return append('starred', len);
    })
    // Catching the AJAX error here is probably too late in most use cases!
    // you're basically saying "if any of these chained stream fail, then
    // start using this single SOME_ERROR_TYPE action stream and then terminate"
    // Effectively, if that AJAX ever throws an exception, this "Epic" will
    // no longer listen for any future events.
    .catch(e => Observable.of({
      type: SOME_ERROR_TYPE,
      payload: e,
      error: true
    }))
);
```

But if we place it inside our `flatMap`, after our `.map()`

```js
const fetchTotalReposLengthEpic = (action$) => (
  action$
    .ofType(FETCH_TOTAL_REPO_STARRED)
    .flatMap(({ username }) => (
      get({
        endpoint: `users/${username}/starred`,
        params: { per_page: 1 },
        resHeader: true
      })
      .map(headers => {
        const reg = /rel="next", <.*&page=(\d+)>; rel="last"/i;
        const len = parseInt(headers.get('link').match(reg)[1], 10);
        return append('starred', len);
      })
      // Now we will prevent any exceptions from propagating to our `ofType()`
      // stream, because our stream inside flatMap is a different chain.
      // so our Epic will continue to listen for future actions, but will
      // flatMap to this provided Observable of one item, the error action.
      // Keeping that in mind, you probably don't want to place any .map()
      // or other transforms *after* this .catch() because they could be acting
      // on this stream of the single SOME_ERROR_TYPE action!
      .catch(e => Observable.of({
        type: SOME_ERROR_TYPE,
        payload: e,
        error: true
      }))
    ))
);
```